metastore/src/java/org/apache/hadoop/hive/metastore/HiveProtoEventsCleanerTask.java

```diff
@@ -0,0 +1,176 @@
+      cleanupDir(basePath);
+    }
+  }
+
+  /**
+   * Returns the expired date partition, using the underlying clock in UTC time.
+   */
+  private static String getExpiredDatePtn() {
+    // Use UTC date to ensure reader date is same on all timezones.
+    LocalDate expiredDate
+            = LocalDateTime.ofEpochSecond((clock.getTime() - ttl) / 1000, 0, ZoneOffset.UTC).toLocalDate();
+    return "date=" + DateTimeFormatter.ISO_LOCAL_DATE.format(expiredDate);
+  }
+
+  /**
+   * Path filters to include only expired date partitions based on TTL.
+   */
+  private static final PathFilter expiredDatePartitionsFilter = new PathFilter() {
+    @Override
+    public boolean accept(Path path) {
+      String dirName = path.getName();
+      return ((dirName.startsWith("date="))
+              && (dirName.compareTo(expiredDatePtn) <= 0));
+    }
+  };
+
+  /**
+   * Finds the expired date partitioned events directory based on TTL and delete them.
+   */
+  private void cleanupDir(Path eventsBasePath) {
+    LOG.debug("Trying to delete expired proto events from " + eventsBasePath);
+    try {
+      FileSystem fs = FileSystem.get(eventsBasePath.toUri(), conf);
+      if (!fs.exists(eventsBasePath)) {
+        return;
+      }
+      FileStatus[] statuses = fs.listStatus(eventsBasePath, expiredDatePartitionsFilter);
+      for (FileStatus dir : statuses) {
+        deleteDir(fs, dir);
+        LOG.info("Deleted expired proto events dir: " + dir.getPath());
+      }
+    } catch (IOException e) {
+      LOG.error("Error while trying to delete expired proto events from " + eventsBasePath, e);
+    }
+  }
+
+  /**
+   * Delete the dir and if it fails, then retry deleting each files with file owner as proxy user.
+   */
+  private void deleteDir(FileSystem fs, FileStatus eventsDir) throws IOException {
+    try {
+      deleteByOwner(fs, eventsDir);
+      return;
+    } catch (IOException e) {
+      // Fall through.
+      LOG.info("Unable to delete the events dir " + eventsDir.getPath()
+              + " and so trying to delete event files one by one.");
+    }
+
+    FileStatus[] statuses = fs.listStatus(eventsDir.getPath());
+    for (FileStatus file : statuses) {
+      deleteByOwner(fs, file);
+    }
+    deleteByOwner(fs, eventsDir);
+  }
+
+  /**
+   * Delete the file/dir with owner as proxy user.
+   */
+  private void deleteByOwner(FileSystem fs, FileStatus fileStatus) throws IOException {
+    String owner = fileStatus.getOwner();
+    if (owner.equals(System.getProperty("user.name"))) {
+      fs.delete(fileStatus.getPath(), true);
+    } else {
+      LOG.info("Deleting " + fileStatus.getPath() + " as user " + owner);
+      UserGroupInformation ugi = UserGroupInformation.createProxyUser(owner,
+              UserGroupInformation.getLoginUser());
+      try {
+        ugi.doAs(new PrivilegedExceptionAction<Object>() {
+          @Override
+          public Object run() throws Exception {
+            fs.delete(fileStatus.getPath(), true);
```

Comment:
FileSystem object is created along with the subject (the user auth) and hence we have to create the FileSystem object in the user context. As a side note, we may have to have a way to cleanup the FileSystem object instances created, one way is to create a FileSystem object and then close it once the file is removed. Another optimization is to run the entire folder in one doAs block, making use of the fact that the folder owner can delete the files in the folder.

Reply:
Fixed. As date partition folder has the permission to delete the files inside, no need to traverse the files. So, number of fs objects will be limited.